packages/adminrouter/extra/src/master/history-service.lua

```diff
@@ -0,0 +1,50 @@
+local function resolve()
+    local mleader = cache.get_cache_entry("mesos_leader")
+    if mleader == nil then
+        ngx.status = ngx.HTTP_SERVICE_UNAVAILABLE
+        ngx.say("503 Service Unavailable: cache is invalid")
+        return ngx.exit(ngx.HTTP_SERVICE_UNAVAILABLE)
+    end
+
+    if mleader['is_local'] == "unknown" then
+        ngx.status = ngx.HTTP_SERVICE_UNAVAILABLE
+        ngx.say("503 Service Unavailable: Mesos leader is unknown.")
+        return ngx.exit(ngx.HTTP_SERVICE_UNAVAILABLE)
+    end
+
+
+    if mleader['is_local'] == 'yes' then
+        -- Let's adjust the URI we send to the upstream service/remove the
+        -- `/dcos-history-service` prefix:
+        ngx.req.set_uri(string.sub(ngx.var.uri, 22))
+        ngx.var.historyservice_upstream = "http://127.0.0.1:15055"
+    else
+        -- Let's prevent infinite proxy loops during failovers when the `leader.mesos`
+        -- can't be reliably determined. Prefixing custom headers with `X-`
+        -- is no longer recommended:
+        -- http://stackoverflow.com/questions/3561381/custom-http-headers-naming-conventions
+        if ngx.req.get_headers()["DCOS-Forwarded"] == "true" then
+            ngx.status = ngx.HTTP_SERVICE_UNAVAILABLE
+            ngx.say("503 Service Unavailable: Mesos leader election in progress")
```

Comment:
Maybe we shouldn't add specific message as we have no idea whether the election is in progress. Remove `Mesos leader election in progress` ?